linera-execution/src/system.rs

```diff
@@ -828,19 +824,27 @@ where
             }
             Withdraw {
                 amount,
-                account:
-                    Account {
-                        owner: Some(owner),
-                        chain_id,
-                    },
+                account: Account { owner, chain_id },
                 user_data: _,
                 recipient,
-            } if chain_id == context.chain_id
-                && context.authenticated_signer.as_ref() == Some(&owner) =>
-            {
-                let balance = self.balances.get_mut_or_default(&owner).await?;
-                if balance.try_sub_assign(amount).is_ok() {
-                    if let Recipient::Account(account) = recipient {
+            } => {
+                ensure!(
+                    chain_id == context.chain_id,
+                    SystemExecutionError::IncorrectChainId(chain_id)
+                );
+                ensure!(
+                    owner.is_some(),
+                    SystemExecutionError::UnauthenticatedClaimOwner
+                );
+                ensure!(
+                    context.authenticated_signer == owner,
+                    SystemExecutionError::UnauthenticatedClaimOwner
```

Comment:
Nit: maybe not now, but should we also distinguish something like `UnauthorizedClaim`, if they're both `Some` but different?

Reply:
So normally, this error would have triggered on the sender chain already. I reused the same error. We could certainly make a pass on system errors and make them more principled.